### ISSUE TITLE
push_count and pop_count overflow

### ISSUE THREAD

**Issue by fovea1959** [EmbedJournal/c-utils#3]
if enough pushes and pops are done, push_count in circular-generic-buffer.c will eventually overflow, causing errors.

Attached test program will show the issue, it eventually gets to 

```
* push_count = 65535, pop_count = 65527
popped 65528
* push_count = 65535, pop_count = 65528
pushed 65536
* push_count =    0, pop_count = 65528
CB is empty
Test Passed, took 0 iteration
* push_count =    0, pop_count = 65528
```

recommend using head and tail like circular_buffer.c, and add a count field to detect buffer full.

Patch:

```
--- a/circular-generic-buffer.h
+++ b/circular-generic-buffer.h
@@ -32,8 +32,9 @@
 
 typedef struct {
     void * const buffer;
-    uint16_t push_count;
-    uint16_t pop_count;
+    uint16_t i_head;
+    uint16_t i_tail;
+    uint16_t count;
     uint16_t const size;
     uint16_t const element_size;
 } circ_gbuf_t;
@@ -42,23 +43,24 @@ typedef struct {
     t x ## _circ_gbuf_data[s];                 \
     circ_gbuf_t x = {                          \
         .buffer = x ## _circ_gbuf_data,        \
-        .push_count = 0,                       \
-        .pop_count = 0,                        \
-        .size = s ,                            \
+        .i_head = 0,                           \
+        .i_tail = 0,                           \
+        .count = 0,                            \
+        .size = s,                             \
         .element_size = sizeof(t)              \
     };

--- a/circular-generic-buffer.c
+++ b/circular-generic-buffer.c
@@ -28,7 +28,7 @@
 
 #include <string.h>
 
-#include "circular-generic-buffer-count.h"
+#include "circular-generic-buffer.h"
 
 static inline void zero_pad(char *s,int len)
 {
@@ -41,11 +41,10 @@ int circ_gbuf_pop(circ_gbuf_t *circ_buf, void *elem, int read_only)
 {
     char *tail;
 
-    if (circ_buf->pop_count >= circ_buf->push_count)
+    if (circ_buf->count == 0)
         return -1;    // quit with an error
 
-    tail = circ_buf->buffer + ((circ_buf->pop_count%circ_buf->size)
-                                * circ_buf->element_size);
+    tail = circ_buf->buffer + (circ_buf->i_tail * circ_buf->element_size);
 
     if (elem)
         memcpy(elem, tail, circ_buf->element_size);
@@ -53,7 +52,8 @@ int circ_gbuf_pop(circ_gbuf_t *circ_buf, void *elem, int read_only)
     if (!read_only) {
         *tail=0;
         zero_pad(tail , circ_buf->element_size);
-        circ_buf->pop_count++;
+        circ_buf->i_tail = (circ_buf->i_tail + 1) % circ_buf->size;
+        circ_buf->count--;
     }
     return 0;
 }
@@ -62,21 +62,32 @@ int circ_gbuf_push(circ_gbuf_t *circ_buf, void *elem)
 {
     char *head;
 
-    if ((circ_buf->push_count - circ_buf->pop_count) >=  circ_buf->size)
+    if (circ_buf->count >=  circ_buf->size)
         return -1;
 
-    head = circ_buf->buffer + ( (circ_buf->push_count % circ_buf->size)
-                                * circ_buf->element_size );
+    head = circ_buf->buffer + (circ_buf->i_head * circ_buf->element_size);
     memcpy(head, elem, circ_buf->element_size);
-    circ_buf->push_count++;
+    circ_buf->i_head = (circ_buf->i_head + 1) % circ_buf->size;
+    circ_buf->count++;
     return 0;
 }
 
 int circ_gbuf_free_space(circ_gbuf_t *circBuf)
 {
-    return circBuf->size - (circBuf->pop_count - circBuf->push_count);
+    return circBuf->size - circBuf->count;
 }
```

[test.zip](https://github.com/EmbedJournal/c-utils/files/2683675/test.zip)

**Comment by SG7** [comment]
Good solution! I have run new code without any problems and it is passing the submitted test.

**Comment by cbsiddharth** [comment]
@fovea1959, very good catch. Thanks for your effort, but the proposed solution isn't thread safe. Please have a look at my solution and how it passes your test case.

The rationale behind using push/pop count as opposed to total element count (as you have done) is to avoid sharing variables across push and pop code paths (threads). In your solution, if you see, `circ_buf->count` is being modified by push and pop. This would mean that this variable has to be locked (when integer access is non-atomic) to guarantee integrity and locking is a pain.

**Comment by SG7** [comment]
@cbsiddharth Are you claiming that your solution is thread safe? That would be great! 

However, I am afraid that simple separation of push and pop variables may not be enough. In multi threaded environment all functions would have to be `mutex` protected.

In `int circ_gbuf_push` function we have many `push_count`  variables.

If many threads are doing the push simultaneously then the `push_count` value may vary from line to line. That would not be good for the calculations.   

Even, if we have only two threads - one producer and one consumer - the simple separation may not be good enough.

For example `circ_gbuf_free_space` could return wrong value if during its execution the `pop_count` or `push_count` was changed by the other thread.

**Comment by cbsiddharth** [comment]
In a lot of embedded use cases, there is one producer (ISRs) and one consumer (application). The above code works well there.

`circ_gbuf_free_space` is read only and no real harm comes from it. I agree it's no super thread safe but it's a step in that direction.

**Comment by fovea1959** [comment]
It is *not* working with single consumer/single producer; that is the test case, and it is failing.,...

    On Wednesday, December 19, 2018, 10:48:19 PM EST, Siddharth Chandrasekaran <notifications@github.com> wrote:  
 
 
In a lot of embedded use cases, there is one producer (ISRs) and one consumer (application). The above code works well there.

circ_gbuf_free_space is read only and no real harm comes from it. I agree it's no super thread safe but it's a step in that direction.

**Comment by SG7** [comment]
@fovea1959 Your test exposed the problem with single consumer and single producer in ONE thread! 

That as you indicated is a problem with the lack of the synchronization regarding wrapping  `push_count` and `pop_count`.    

In the worst case the problem may not be fixable by the way how push and pop operates now. One cannot run two separate entities operating in totally independent way over the same container.  Push and pop has to know more about themselves.

**Comment by cbsiddharth** [comment]
@fovea1959, Yes. I see you mention it's _not_ working. That's why I reopened this issue. I. Will get to it ASAP.

The explanation was for @SG7 on his comment on thread safety, which is valid with or without the issue you mention.

**Comment by cbsiddharth** [comment]
@SG7, don't loose hope just yet. It can be done. This was a work in progress that got interrupted. I will finish this soon.

**Comment by cbsiddharth** [comment]
@fovea1959, this issue is resolved. I have added a new tests directory to have more thorough testing. I have also added your test case into `test/test-circular-buffer.c` under `test_probabilistic()`.
